File: leet-code/dynamic_programming/unique_paths.rb
def unique_paths(grid)
  # if the first cell is an obstacle immediately return 0 (no possible path)
  return 0 if grid[0][0] == 1

  # set up the scene
  row_last_index = grid[0].length - 1
  column_last_index = grid.length - 1
  grid[0][0] = 1

  # fill in the first row
  1.upto(row_last_index) do |idx|
    grid[0][idx] == 1 ? grid[0][idx] = 0 : grid[0][idx] = grid[0][idx -1 ]
  end

  # fill in the first column
  1.upto(column_last_index) do |idx|
    grid[idx][0] == 1 ? grid[idx][0] = 0 : grid[idx][0] = grid[idx - 1][0]
  end

  # fill in the rest of the matrix based on previous values
  1.upto(column_last_index) do |x|
    1.upto(row_last_index) do |y|
      if grid[x][y] == 1
        grid[x][y] = 0
      else
        grid[x][y] = grid[x - 1][y] + grid[x][y - 1]
      end
    end
  end

  grid[column_last_index][row_last_index]
end

File: leet-code/dynamic_programming/unique_paths.js
function uniquePaths(grid) {
  // if the first cell is an obstacle immediately return 0 (no possible path)
  if (grid[0][0] === 1) return 0;

  // set up the scene
  grid[0][0] = 1;

  // fill in the first row
  let rowLength = grid[0].length;

  for (let i = 1; i < rowLength; i += 1) {
    if (grid[0][i] === 1) {
      grid[0][i] = 0;
    } else {
      grid[0][i] = grid[0][i - 1];
    }
  }

  // fill in the first column
  let columnLength = grid.length;

  for (let i = 1; i < columnLength; i += 1) {
    if (grid[i][0] === 1) {
      grid[i][0] = 0;
    } else {
      grid[i][0] = grid[i - 1][0];
    }
  }

  // fill in the rest of the matrix based on previous values
  for (let i = 1; i < columnLength; i += 1) {
    for (let j = 1; j < rowLength; j += 1) {
      if (grid[i][j] === 1) {
        grid[i][j] = 0;
      } else {
        grid[i][j] = grid[i - 1][j] + grid[i][j - 1];
      }
    }
  }

  // return value in thebottom right cell
  return grid[columnLength - 1][rowLength - 1];
}

let grid = [[0, 0, 0], [0, 1, 0], [0, 0, 0]];
console.log(uniquePaths(grid));
